File: Lab_7/quiz-server/src/main/java/com/example/quiz/Quiz.java
package com.example.quiz;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Represents a quiz with multiple questions and tracking for user answers.
 */
public class Quiz {
    private String quizId;
    private String userName;
    private List<Question> questions;
    private Map<Integer, Integer> userAnswers; // questionId -> answerIndex
    private int currentQuestionIndex;
    private boolean completed;

    public Quiz() {
        this.questions = new ArrayList<>();
        this.userAnswers = new HashMap<>();
        this.currentQuestionIndex = 0;
        this.completed = false;
    }

    public Quiz(String quizId, String userName, List<Question> questions) {
        this.quizId = quizId;
        this.userName = userName;
        this.questions = questions;
        this.userAnswers = new HashMap<>();
        this.currentQuestionIndex = 0;
        this.completed = false;
    }

    // Getters and Setters
    public String getQuizId() {
        return quizId;
    }

    public void setQuizId(String quizId) {
        this.quizId = quizId;
    }

    public String getUserName() {
        return userName;
    }

    public void setUserName(String userName) {
        this.userName = userName;
    }

    public List<Question> getQuestions() {
        return questions;
    }

    public void setQuestions(List<Question> questions) {
        this.questions = questions;
    }

    public Map<Integer, Integer> getUserAnswers() {
        return userAnswers;
    }

    public void setUserAnswers(Map<Integer, Integer> userAnswers) {
        this.userAnswers = userAnswers;
    }

    public int getCurrentQuestionIndex() {
        return currentQuestionIndex;
    }

    public void setCurrentQuestionIndex(int currentQuestionIndex) {
        this.currentQuestionIndex = currentQuestionIndex;
    }

    public boolean isCompleted() {
        return completed;
    }

    public void setCompleted(boolean completed) {
        this.completed = completed;
    }

    /**
     * Get the current question
     * @return the current Question or null if no more questions
     */
    public Question getCurrentQuestion() {
        if (currentQuestionIndex >= 0 && currentQuestionIndex < questions.size()) {
            return questions.get(currentQuestionIndex);
        }
        return null;
    }

    /**
     * Submit an answer for a question
     * @param questionId the ID of the question
     * @param answerIndex the index of the selected answer
     */
    public void submitAnswer(int questionId, int answerIndex) {
        userAnswers.put(questionId, answerIndex);
    }

    /**
     * Move to the next question
     * @return true if moved to next question, false if quiz is complete
     */
    public boolean nextQuestion() {
        if (currentQuestionIndex < questions.size() - 1) {
            currentQuestionIndex++;
            return true;
        } else {
            completed = true;
            return false;
        }
    }

    /**
     * Calculate the score based on correct answers
     * @return the number of correct answers
     */
    public int calculateScore() {
        int score = 0;
        for (Question question : questions) {
            Integer userAnswer = userAnswers.get(question.getId());
            if (userAnswer != null && question.isCorrect(userAnswer)) {
                score++;
            }
        }
        return score;
    }

    /**
     * Get the total number of questions
     * @return the total number of questions
     */
    public int getTotalQuestions() {
        return questions.size();
    }

    /**
     * Get the percentage score
     * @return the percentage score (0-100)
     */
    public double getPercentageScore() {
        if (questions.isEmpty()) {
            return 0.0;
        }
        return (calculateScore() * 100.0) / questions.size();
    }
}

File: Lab_8/src/models/Quiz.js
/**
 * Quiz Model
 * Manages quiz state, questions, and user answers
 */
export class Quiz {
  constructor(questions = []) {
    this.questions = questions;
    this.currentQuestionIndex = 0;
    this.userAnswers = new Map();
    this.completed = false;
    this.startTime = Date.now();
    this.endTime = null;
  }

  /**
   * Get the current question
   * @returns {Question|null} Current question or null if none available
   */
  getCurrentQuestion() {
    if (this.currentQuestionIndex < this.questions.length) {
      return this.questions[this.currentQuestionIndex];
    }
    return null;
  }

  /**
   * Submit an answer for the current question
   * @param {number} answerIndex - The selected answer index
   */
  submitAnswer(answerIndex) {
    const currentQuestion = this.getCurrentQuestion();
    if (currentQuestion) {
      this.userAnswers.set(currentQuestion.id, answerIndex);
    }
  }

  /**
   * Move to the next question
   * @returns {boolean} True if moved to next, false if quiz is complete
   */
  nextQuestion() {
    if (this.currentQuestionIndex < this.questions.length - 1) {
      this.currentQuestionIndex++;
      return true;
    }
    this.completed = true;
    this.endTime = Date.now();
    return false;
  }

  /**
   * Move to the previous question
   * @returns {boolean} True if moved to previous, false if at start
   */
  previousQuestion() {
    if (this.currentQuestionIndex > 0) {
      this.currentQuestionIndex--;
      return true;
    }
    return false;
  }

  /**
   * Calculate the total score
   * @returns {number} Number of correct answers
   */
  calculateScore() {
    let score = 0;
    this.questions.forEach(question => {
      const userAnswer = this.userAnswers.get(question.id);
      if (userAnswer !== undefined && question.isCorrect(userAnswer)) {
        score++;
      }
    });
    return score;
  }

  /**
   * Get the percentage score
   * @returns {number} Percentage score (0-100)
   */
  getPercentageScore() {
    if (this.questions.length === 0) return 0;
    return (this.calculateScore() / this.questions.length) * 100;
  }

  /**
   * Get total number of questions
   * @returns {number} Total questions
   */
  getTotalQuestions() {
    return this.questions.length;
  }

  /**
   * Check if a question has been answered
   * @param {number} questionId - The question ID
   * @returns {boolean} True if answered
   */
  isQuestionAnswered(questionId) {
    return this.userAnswers.has(questionId);
  }

  /**
   * Get the user's answer for a question
   * @param {number} questionId - The question ID
   * @returns {number|undefined} The answer index or undefined
   */
  getUserAnswer(questionId) {
    return this.userAnswers.get(questionId);
  }

  /**
   * Get quiz duration in seconds
   * @returns {number} Duration in seconds
   */
  getDuration() {
    const end = this.endTime || Date.now();
    return Math.floor((end - this.startTime) / 1000);
  }

  /**
   * Get detailed results for each question
   * @returns {Array} Array of result objects
   */
  getDetailedResults() {
    return this.questions.map(question => {
      const userAnswer = this.userAnswers.get(question.id);
      const isCorrect = userAnswer !== undefined && question.isCorrect(userAnswer);
      
      return {
        questionId: question.id,
        questionText: question.text,
        category: question.category,
        userAnswer: userAnswer !== undefined ? question.options[userAnswer] : 'Not answered',
        correctAnswer: question.getCorrectAnswer(),
        isCorrect: isCorrect,
        answered: userAnswer !== undefined
      };
    });
  }

  /**
   * Reset the quiz to start over
   */
  reset() {
    this.currentQuestionIndex = 0;
    this.userAnswers.clear();
    this.completed = false;
    this.startTime = Date.now();
    this.endTime = null;
  }
}

export default Quiz;
